rosbag2_test_common/include/rosbag2_test_common/temporary_directory_fixture.hpp

```diff
@@ -17,26 +17,72 @@
 
 #include <gmock/gmock.h>
 
+#include <cstdio>
+#include <exception>
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+#include <random>
+#include <sstream>
 #include <string>
 
-#include "rcpputils/filesystem_helper.hpp"
-
 using namespace ::testing;  // NOLINT
 
 namespace rosbag2_test_common
 {
 
+/// \brief Construct a uniquely named temporary directory, in "parent", with format base_nameXXXXXX
+/// The output, if successful, is guaranteed to be a newly-created directory.
+/// The underlying implementation keeps generating paths until one that does not exist is found or
+/// until the number of iterations exceeded the maximum tries.
+/// This guarantees that there will be no existing files in the returned directory.
+/// \param[in] base_name User-specified portion of the created directory.
+/// \param[in] parent_path The parent path of the directory that will be created.
+/// \param[in] max_tries The maximum number of tries to find a unique directory (default 1000)
+/// \return A path to a newly-created directory with base_name and a 6-character unique suffix.
+/// \throws std::system_error If any OS APIs do not succeed.
+/// \throws std::runtime_error If the number of the iterations exceeds the maximum tries and
+/// a unique directory is not found.
+std::filesystem::path create_temp_directory(
```

Comment:
why you don't use `rcpputils::fs::create_temp_directory` ?

Reply:
Yeah, I think we should switch to using the `rcpputils::fs::create_temp_directory` here.  If there are still problems with it, then we should fix them over in `rcpputils`.

Reply:
@ahcorde @clalancette The reasons why I wrote my own implementation of the `create_temp_directory` is the following:
1. Because [last time] when we tried to use `rcpputils::fs::create_temp_directory` alongside with cleanup via `std::filesystem::remove_all`, the `std::filesystem::remove_all` was failing on Windows and also caused some test failures. See https://github.com/ros2/rosbag2/pull/1576#issuecomment-2006060701 and https://github.com/ros2/rosbag2/pull/1576#pullrequestreview-1950086019 for details. 
I have tried to triage this problem on Windows. It turns out that in failing tests the rosbag2 creating subfolder with files in the temporary created folder from `rcpputils::fs::create_temp_directory`. For some reason, likely because `rcpputils::fs::create_temp_directory` inside using platform-dependent function calls, i.e., `_mktemp_s(..)` and own implementation for `create_directories(const path & p)` the `std::filesystem::remove_all` fails when trying to delete such folder structures. However, if create a temp folder manually or use the temp folder returned by the `std::filesystem::temp_directory_path()` without creating a subfolder in it with `rcpputils::fs::create_temp_directory`, the `std::filesystem::remove_all` is not failing.
2. As far as I understand, it was a motion to deprecate functions from the `rcpputils::fs` and use alternatives from the `std::filesystem`.
3. I don't like the idea of usage `std::filesystem::path` with the `rcpputils::fs::path`. It would be more consistent and more robust to use only `std::filesystem::path`. If we decided to get rid from the `rcpputils::fs` let's do it! We are using  `rcpputils::fs::create_temp_directory` only in two places in the core packages in the `rosbag2` and in tests for the [rcl_logging_spdlog](https://github.com/ros2/rcl_logging/blob/rolling/rcl_logging_spdlog/test/test_logging_interface.cpp#L110).

Let's see how Windows CI will pass. If it will pass without failures, I don't mind to move my `rosbag2_test_common::create_temp_directory(..)` to the `rcpputils::fs` namespace.

Reply:
> * I don't like the idea of usage `std::filesystem::path` with the `rcpputils::fs::path`. It would be more consistent and more robust to use only `std::filesystem::path`. If we decided to get rid from the `rcpputils::fs` let's do it! We are using  `rcpputils::fs::create_temp_directory` only in two places in the core packages in the `rosbag2` and in tests for the [rcl_logging_spdlog](https://github.com/ros2/rcl_logging/blob/rolling/rcl_logging_spdlog/test/test_logging_interface.cpp#L110).

Yes, we are actively doing this right now in https://github.com/ros2/rcpputils/pull/197 and https://github.com/ros2/rcpputils/pull/196 .  That said, `std::filesystem` doesn't have any equivalent to `rcpputils::fs::create_temp_directory`, so we are going to have to carry around that function indefinitely.

Reply:
> @clalancette If CI will pass green I would like to move my implementation of the `create_temp_directory` to the `rcpputils::fs` as is but with the new name `create_temporary_directory` and deprecate the original `rcpputils::fs::create_temp_directory` eventually.

I don't think we should do that.  We are currently in the process of breaking the API for `create_temp_directory` anyway, so I think we should do this all at the same time.  Please take a look at https://github.com/ros2/rcpputils/pull/197 and give feedback on what we think the API should be.

Reply:
BTW. The implementation in the https://github.com/ros2/rcpputils/pull/197/files has API-breaking changes and still uses platform-dependent code, which is not good.  
Creating a new function with a different name and deprecating the old one would be better.

Reply:
> BTW. The implementation in the https://github.com/ros2/rcpputils/pull/197/files has API-breaking changes and still uses platform-dependent code, which is not good.

Yes, it has API-breaking changes.  See the discussion in https://github.com/ros2/rcpputils/pull/196#pullrequestreview-2167477783 about why we decided to do that.

Reply:
@clalancette I see. However, I still don't like the idea of doing API-breaking changes, which is require to merge 3 PRs in synch.  
Why not do it sequentially by creating a new function with `create_temporary_directory` name? 
I've tried to look out, and it seems this name is not occupied in the ROS 2 core packages.

Reply:
@clalancette FYI In Apex.AI, when we are doing synch with upstream ROS 2 by pulling PRs each time, such breaking changes, which require merging multiple PRs in different packages in synch, is a nightmare since it is breaking our CI.

We would be grateful if such breaking changes will be as minimum as possible.